### PR TITLE
refactor: tr_quickfindFirstK --> std::partial_sort

### PR DESCRIPTION
**Pull request by ckerr**
Remove `tr_quickfindFirstK()` and use `std::partial_sort()` instead.

### CODE REVIEW

File: libtransmission/announcer.cc [hunk 1]
```diff
@@ -1567,17 +1569,15 @@ static void scrape_request_delegate(
...
+static void multiscrape(tr_announcer* announcer, std::vector<tr_tier*> const& tiers)
 {
     size_t request_count = 0;
     time_t const now = tr_time();
-    size_t const tier_count = tr_ptrArraySize(tiers);
     tr_scrape_request requests[MAX_SCRAPES_PER_UPKEEP] = {};
 
     /* batch as many info_hashes into a request as we can */
-    for (size_t i = 0; i < tier_count; ++i)
+    for (auto& tier : tiers)
```
Comment (mikedld) [inline libtransmission/announcer.cc:1579]: Here and elsewhere: if something is a pointer, I'd like to be explicit about it.
```suggestion
    for (auto* tier : tiers)
```

File: libtransmission/announcer.cc [hunk 2]
```diff
@@ -1741,22 +1738,25 @@ static void scrapeAndAnnounceMore(tr_announcer* announcer)
...
+    {
+        std::partial_sort(
+            std::begin(announce_me),
+            std::begin(announce_me) + MAX_ANNOUNCES_PER_UPKEEP,
+            std::end(announce_me),
+            [](tr_tier const* a, tr_tier const* b) { return compareAnnounceTiers(a, b) < 0; });
+        announce_me.resize(MAX_ANNOUNCES_PER_UPKEEP);
+    }
+
+    for (auto& tier : announce_me)
```
Comment (mikedld) [inline libtransmission/announcer.cc:1755]:
```suggestion
    for (auto* tier : announce_me)
```

File: libtransmission/announcer.cc [hunk 3]
```diff
@@ -1741,22 +1738,25 @@ static void scrapeAndAnnounceMore(tr_announcer* announcer)
...
      * available, use compareAnnounceTiers to prioritize. */
-    int n = MIN(tr_ptrArraySize(&announceMe), MAX_ANNOUNCES_PER_UPKEEP);
-    for (int i = 0; i < n; ++i)
+    if (announce_me.size() > MAX_ANNOUNCES_PER_UPKEEP)
+    {
+        std::partial_sort(
+            std::begin(announce_me),
+            std::begin(announce_me) + MAX_ANNOUNCES_PER_UPKEEP,
+            std::end(announce_me),
+            [](tr_tier const* a, tr_tier const* b) { return compareAnnounceTiers(a, b) < 0; });
```
Comment (mikedld) [inline libtransmission/announcer.cc:1751]: That's the only place where `compareAnnounceTiers()` is used, maybe return `bool` (if less) from it right away; then lambda won't even be needed.

File: libtransmission/session.cc [hunk 4]
```diff
@@ -3017,85 +3019,40 @@ int tr_sessionGetAntiBruteForceThreshold(tr_session const* session)
...
-            candidates,
-            num_candidates,
-            sizeof(struct TorrentAndPosition),
-            compareTorrentAndPositions,
-            num_wanted);
+        std::partial_sort(
+            std::begin(candidates),
+            std::begin(candidates) + num_wanted,
+            std::end(candidates),
+            [](tr_torrent* a, tr_torrent* b) { return tr_torrentGetQueuePosition(a) < tr_torrentGetQueuePosition(b); });
```
Comment (mikedld) [inline libtransmission/session.cc:3047]:
```suggestion
            [](auto const* a, auto const* b) { return tr_torrentGetQueuePosition(a) < tr_torrentGetQueuePosition(b); });
```

File: libtransmission/peer-mgr.cc [hunk 5]
```diff
@@ -4337,11 +4309,15 @@ static struct peer_candidate* getPeerCandidates(tr_session* session, int* candid
...
+    auto const n_candidates = walk - candidates;
+    *candidateCount = n_candidates;
+    if (n_candidates > max)
     {
-        selectPeerCandidates(candidates, walk - candidates, max);
+        std::partial_sort(
+            candidates,
+            candidates + max,
+            candidates + n_candidates,
+            [](peer_candidate const& a, peer_candidate const& b) { return a.score < b.score; });
```
Comment (mikedld) [inline libtransmission/peer-mgr.cc:4320]:
```suggestion
            [](auto const& a, auto const& b) { return a.score < b.score; });
```

File: libtransmission/announcer.cc (hunk 3 above)
Comment (mikedld) [reply libtransmission/announcer.cc:1751]: Here and elsewhere: spelling out parameter type isn't necessary (unless you want to).
```suggestion
            [](auto const* a, auto const* b) { return compareAnnounceTiers(a, b) < 0; });
```

File: libtransmission/session.cc [hunk 6]
```diff
@@ -3017,85 +3019,40 @@ int tr_sessionGetAntiBruteForceThreshold(tr_session const* session)
...
+        candidates.resize(num_wanted);
     }
 
-    /* add them to the return array */
-    for (size_t i = 0; i < num_wanted; ++i)
+    // add them to the return array
+    for (auto const& candidate : candidates)
```
Comment (mikedld) [inline libtransmission/session.cc:3052]:
```suggestion
    for (auto const* candidate : candidates)
```

File: libtransmission/announcer.cc [hunk 7]
```diff
@@ -6,10 +6,12 @@
  *
  */
 
+#include <algorithm> // std::partial_sort
```
Comment (mikedld) [inline libtransmission/announcer.cc:9]: Do we maybe want to put standard C++ headers into a separate group?

**Review by mikedld — APPROVED** [review]

File: libtransmission/announcer.cc (hunk 1 above)
Comment (ckerr) [reply libtransmission/announcer.cc]: fixed ecd14048b7f4a2741646eb10d54d8064a902585a

File: libtransmission/announcer.cc (hunk 2 above)
Comment (ckerr) [reply libtransmission/announcer.cc]: fixed ecd14048b7f4a2741646eb10d54d8064a902585a

File: libtransmission/announcer.cc (hunk 3 above)
Comment (ckerr) [reply libtransmission/announcer.cc]: I had made it explicit out of fear that the older compilers in CI wouldn't handle it (e.g. gcc 5.5.0), but you're right. Let's give it a try.

Fixed ecd14048b7f4a2741646eb10d54d8064a902585a

File: libtransmission/session.cc (hunk 4 above)
Comment (ckerr) [reply libtransmission/session.cc]: ecd14048b7f4a2741646eb10d54d8064a902585a

File: libtransmission/peer-mgr.cc (hunk 5 above)
Comment (ckerr) [reply libtransmission/peer-mgr.cc]: ecd14048b7f4a2741646eb10d54d8064a902585a

File: libtransmission/session.cc (hunk 6 above)
Comment (ckerr) [reply libtransmission/session.cc]: ecd14048b7f4a2741646eb10d54d8064a902585a

**Pull request merged**